### DISCUSSION TITLE
transfer_failed with previous approve and message sign if there are 2 tokens no signed

### DISCUSSION THREAD

**Discussion by facilficom** [Elnaril/uniswap-universal-router-decoder#70] [Q&A]
When I attempt to swap token A for token B and neither token A nor token B has been approved and signed message, the transaction fails.  the fail give me transfer_failed but its only for the estimate gas that cannot be calculated in this case (doesnt work if i put a manual higher gas too)
However, when I first swap token A for WETH, and then swap token B for WETH in other transactions with sign message and approve,  the previous failed one, now succeeds.
...

            allowance_amount = feg_token_contract.functions.allowance(
                from_address_checksum,
                permit2_address_checksum,
            ).call()

            if allowance_amount < total_amount:
                permit2_allowance = int(2**256 - 1)
                contract_function = feg_token_contract.functions.approve(
                    permit2_address_checksum,
                    permit2_allowance
                )
                trx_params = contract_function.build_transaction({
                        'from': from_address_checksum,
                        'chainId': chain_id,     
                        'nonce': nonce,
                        'gasPrice': gas_price                    
                })
                trx_params_gas_estimate = web3.eth.estimate_gas(trx_params)
                trx_params.update({'gas': int(trx_params_gas_estimate * 1.5)})
                signed_transaction = web3.eth.account.sign_transaction(trx_params, private_key)
                trx_hash = web3.eth.send_raw_transaction(signed_transaction.raw_transaction)
                receipt = web3.eth.wait_for_transaction_receipt(trx_hash)
                trx_hash_hex = web3.to_hex(trx_hash)
                trx_link = get_explorer_url(current_network, trx_hash_hex)
                if receipt['status'] == 0:
                    return f"Approve Transaction Failed!\nTX Hash: {trx_hash_hex}\nTX Link: {trx_link}"
                else:
                    nonce = get_nonce(web3, from_address_checksum)

            chain_builder = codec.encode.chain()

            permit2_contract = web3.eth.contract(address=permit2_address_checksum, abi=PERMIT2_ABI)

            p2_amount, p2_expiration, p2_nonce = permit2_contract.functions.allowance(
                from_address_checksum,
                checksum_address_feg,
                ur_address_checksum
            ).call()
            current_time = int(datetime.now().timestamp())

            if p2_amount < total_amount or p2_expiration < current_time:
                allowance_amount = int(2**160 - 1)
                permit_data, signable_message = codec.create_permit2_signable_message(
                    checksum_address_feg,
                    allowance_amount,
                    codec.get_default_expiration(),
                    p2_nonce,
                    ur_address_checksum,
                    codec.get_default_deadline(),
                    chain_id,
                    permit2_address_checksum
                )
                signed_message = account.sign_message(signable_message)
                chain_builder = chain_builder.permit2_permit(permit_data, signed_message)

            path = [checksum_address_feg, checksum_address_token]
            commission_wei = int(total_amount * COMMISSION_RATE)
            amount_without_commission = total_amount - commission_wei
            path2 = [checksum_address_feg, checksum_address_weth]

            if slippage == 0:
                amount_out_min = 0
            else:
                output = uniswap_contract.functions.getAmountsOut(amount_without_commission, path).call()
                amount_out_min = int(output[1] * (1 - (slippage / 100)))

            chain_builder = chain_builder.v2_swap_exact_in(FunctionRecipient.SENDER, amount_without_commission, amount_out_min, path, payer_is_sender=True)
            chain_builder = chain_builder.v2_swap_exact_in(FunctionRecipient.ROUTER, commission_wei, 0, path2, payer_is_sender=True)
            chain_builder = chain_builder.unwrap_weth(FunctionRecipient.CUSTOM, 0, owner_wallet_address_checksum)     

            encoded_input = chain_builder.build(codec.get_default_deadline())

            trx_params = {
                "from": from_address_checksum,
                "to": ur_address_checksum,
                "maxPriorityFeePerGas": maxpriorityfeepergas,
                "maxFeePerGas": maxfeepergas,
                "type": '0x2',
                "chainId": chain_id,
                "value": 0,
                "nonce": nonce,
                "data": encoded_input,
            }
            trx_params_gas_estimate = web3.eth.estimate_gas(trx_params)
            trx_params.update({"gas": int(trx_params_gas_estimate * 1.5)})
            signed_transaction = web3.eth.account.sign_transaction(trx_params, private_key)
            trx_hash = web3.eth.send_raw_transaction(signed_transaction.raw_transaction)
            receipt = web3.eth.wait_for_transaction_receipt(trx_hash)

**Comment by Elnaril** [comment]
Hello @facilficom 
Not sure to follow you, but here is my understanding of what you're doing:

1/ You create a swap transaction of a token that you did not approve (token + permit2) and the gas estimation fails.
=> This is normal: to compute the gas estimation, the node tries to execute the transactions locally which fails because of the missing approval

2/ Then you approve the token (token + permit2) for another transaction and this swap succeeds. Then you estimate again the previous transaction and this time it works.
=> This is also normal, under the condition you approved enough (amount and time) for both transactions.

Does it answer your question?

**Reply by facilficom** [reply]
Yes, you’ve understood it correctly. My main issue was with the gas estimation failing initially due to the lack of token approval + permit2 when swap tokenA for tokenB. do you know other method to estimate gas so i can try it and reply to you if this was the problem?

**Reply by Elnaril** [reply]
There is not really another method to estimate the gas needed by a transaction ...
Though you can approve the token, and estimate and/or execute the transaction on a local fork if that can help to solve your issue.

**Reply by facilficom** [reply]
https://bscscan.com/address/0x9a205aF2FA62bF5148533d93Eaad95EaD66453f8 i have approve and permit2 but nothing, doesnt work and the gas is estimate all is correct, sometimes work with some wallet and i dont know why and other times no. im trying sdoge sell for feg. really dont understand why. i will write some error: An unexpected error occurred: Can't parse entities: can't find end of the entity starting at byte offset 67. An unexpected error occurred: Can't parse entities: can't find end of the entity starting at byte offset 58. An unexpected error occurred:
Transaction HexBytes('0x5ee68c1ce709137be6f494bb948524e373765ee94298e71c57df02534e6dc6bf') is not in the chain after 120 seconds. and other times is totally blocked at this function

**Reply by Elnaril** [reply]
```
An unexpected error occurred: Can't parse entities: can't find end of the entity starting at byte offset 67.
An unexpected error occurred: Can't parse entities: can't find end of the entity starting at byte offset 58
```
Sounds like TG errors to me, not web3 or UR SDK ones ...

-----

```
Transaction HexBytes('0x5ee68c1ce709137be6f494bb948524e373765ee94298e71c57df02534e6dc6bf') is not in the chain after 120 seconds
```
Did you set `maxPriorityFeePerGas` and `maxFeePerGas` high enough given the blockchain load at that time?
If yes, can you please post the `trx_params` of such failing transaction ?

**Comment by facilficom** [comment]
current_network = context.user_data.get('network')
        web3 = get_web3(current_network)
        web3_protect = get_web3_protect(current_network)
        codec = RouterCodec(web3)
        maxpriorityfeepergas, maxfeepergas = compute_gas_fees(web3_protect)
        gas_price = int(web3.eth.gas_price * 1.2)
        minpriorityfeepergas = web3.to_wei(1, 'gwei')
        maxpriorityfeepergas = max(maxpriorityfeepergas, minpriorityfeepergas)
        if maxfeepergas < gas_price + maxpriorityfeepergas:
            maxfeepergas = int((gas_price + maxpriorityfeepergas) * 1.2)
        current_addresses = get_addresses(current_network)
        chain_id = web3.eth.chain_id
this is what i did
same situation, same amount, turn on the bot and call the function 2 times, wihtout change nothing:

{'from': '0x9a205aF2FA62bF5148533d93Eaad95EaD66453f8', 'to': '0x1A0A18AC4BECDDbd6389559687d1A73d8927E416', 'maxPriorityFeePerGas': 3750000000, 'maxFeePerGas': 
5940000000, 'type': '0x2', 'chainId': 56, 'value': 0, 'nonce': 12, 'data': '0x3593564c000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000066c0560000000000000000000000000000000000000000000000000000000000000000040806080c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000001a00000000000000000000000000000000000000000000000000000000000000220000000000000000000000000000000000000000000000000000000000000034000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000001158e460913d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c3a71cb66e39af44d3c7dbdf00b24d16d1da4d4b000000000000000000000000f3c7cecf8cbc3066f9a87b310cebe198d00479ac0000000000000000000000000000000000000000000000000000000000000060000000000000000000000000f3c7cecf8cbc3066f9a87b310cebe198d00479ac00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000002648000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000028000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002000000000000000000000000f3c7cecf8cbc3066f9a87b310cebe198d00479ac000000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c00000000000000000000000000000000000000000000000000000000000000400000000000000000000000007724538ad41ff70a1f27a8b22b25157ca61037d10000000000000000000000000000000000000000000000000000000000000000'} with this An unexpected error occurred: Can't parse entities: can't find end of the entity starting at byte offset 67....


{'from': '0x9a205aF2FA62bF5148533d93Eaad95EaD66453f8', 'to': '0x1A0A18AC4BECDDbd6389559687d1A73d8927E416', 'maxPriorityFeePerGas': 1265656250, 'maxFeePerGas': 
2958787500, 'type': '0x2', 'chainId': 56, 'value': 0, 'nonce': 12, 'data': '0x3593564c000000000000000000000000000000000000000000000000000000000000006000000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000066c0563200000000000000000000000000000000000000000000000000000000000000040806080c000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000004000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000001a00000000000000000000000000000000000000000000000000000000000000220000000000000000000000000000000000000000000000000000000000000034000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000001158e460913d00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000002000000000000000000000000c3a71cb66e39af44d3c7dbdf00b24d16d1da4d4b000000000000000000000000f3c7cecf8cbc3066f9a87b310cebe198d00479ac0000000000000000000000000000000000000000000000000000000000000060000000000000000000000000f3c7cecf8cbc3066f9a87b310cebe198d00479ac00000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000002648000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000000028000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000a000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002000000000000000000000000f3c7cecf8cbc3066f9a87b310cebe198d00479ac000000000000000000000000bb4cdb9cbd36b01bd1cbaebf2de08d9173bc095c00000000000000000000000000000000000000000000000000000000000000400000000000000000000000007724538ad41ff70a1f27a8b22b25157ca61037d10000000000000000000000000000000000000000000000000000000000000000'}
with this An unexpected error occurred:
Transaction HexBytes('0x571668fd4823a72ac62ff1684b86d4dc253e15212b150f540d2f6e186475e411') is not in the chain after 120 seconds


real history:

Please enter the amount of sDOGE to sell in the format: `amount`

20

An unexpected error occurred: Can't parse entities: can't find end of the entity starting at byte offset 67


Please enter the amount of sDOGE to sell in the format: `amount`

20


An unexpected error occurred:
Transaction HexBytes('0x571668fd4823a72ac62ff1684b86d4dc253e15212b150f540d2f6e186475e411') is not in the chain after 120 seconds

if i change wallet, or buy them with weth. the transaction work then (previous tryies)

**Reply by Elnaril** [reply]
Are you sure of your nonce? You have currently 16 transactions from this wallet (`0x9a205aF2FA62bF5148533d93Eaad95EaD66453f8`), but your nonce is 12. If it's an old transaction and your nonce was correct, that could mean you still have your transaction in the mempool. In this case, you need to override it with a bigger gas fee.

**Reply by playsmartdefi** [reply]
This is other account sorry im not in pc, the transactions were of 3 Hours ago, when I write to you. If you open the last transaction from the link of bscscan in the part 'state' you can see nonce after last transaction: 12

**Reply by Elnaril** [reply]
Yes, indeed. In this case, _if the same transaction works from another wallet_, the only issue I can think of is that you already have a pending transaction with `nonce` = 12. Either you wait an undefined period until it's cleared from the mempool, or you override it: same nonce = 12, but higher gas fees (used to be +12.5%, not sure nowadays)

**Reply by playsmartdefi** [reply]
Nothing to do, i try with all possible solution, sometimes the gas was not estimate for not have tx estimate balance, was more than real need it. But to swap token vs token (no weth) the transaction doesnt go sometimes so for this i come back to call directly on router v2. When I wil integrate v3 too i will try if there goes the tx

**Reply by facilficom** [reply]
maybe need to check if the pernit message's data is correct when try to swap 2 tokens without bnb pair? i seen different data between my logs and logs from pancakeswap  with same transaction when i use it for bnb pair
